aten/src/ATen/native/sparse/cuda/SoftMax.cu

```diff
@@ -0,0 +1,645 @@
+            auto indices_row = indices_accessor[j];
+            auto stride = strides_ptr[j];
+            pool_index += stride * indices_row[x];
+          }
+        }
+        return pool_index;
+      });
+  int64_t* pools_ptr = pools.data().get();
+
+  thrust::device_vector<int64_t> sorted_indices(nnz);
+  thrust::sequence(sorted_indices.begin(), sorted_indices.end(), 0);
+
+  thrust::sort(
+      sorted_indices.begin(),
+      sorted_indices.end(),
+      [pools_ptr] __device__(int64_t x, int64_t y) {
+        return pools_ptr[x] < pools_ptr[y];
+      });
+  thrust::device_vector<int64_t> count_set(nnz);
+  thrust::device_vector<scalar_t*> mx_rows;
+  thrust::device_vector<scalar_t> log_softmax_max_buffer;
+  if (requireMxRows) {
+    thrust::device_vector<int64_t> mx_rows_indices(nnz);
+    auto values_accessor =
+        values.packed_accessor<scalar_t, 2>(); // {nnz, nvalues}
+
+    auto val_zip_ter = thrust::make_zip_iterator(thrust::make_tuple(
+        sorted_indices.begin(), thrust::make_constant_iterator(int64_t(1))));
+    auto out_val_zip_ter = thrust::make_zip_iterator(
+        thrust::make_tuple(mx_rows_indices.begin(), count_set.begin()));
+
+    auto new_end = thrust::reduce_by_key(
+        sorted_indices.begin(),
+        sorted_indices.end(),
+        val_zip_ter,
+        thrust::make_discard_iterator(),
+        out_val_zip_ter,
+        [pools_ptr] __device__(int64_t x, int64_t y) {
+          return pools_ptr[x] == pools_ptr[y];
+        },
+        [nvalues, values_accessor] __device__(
+            thrust::tuple<int64_t, int64_t> x,
+            thrust::tuple<int64_t, int64_t> y) {
+          auto index_x = thrust::get<0>(x);
+          auto count_x = thrust::get<1>(x);
+
+          auto index_y = thrust::get<0>(y);
+          auto count_y = thrust::get<1>(y);
+          const scalar_t* a = values_accessor[index_x].data();
+          const scalar_t* b = values_accessor[index_y].data();
+
+          int64_t j = 0;
+          while (j < nvalues && a[j] > b[j]) {
+            j++;
+          }
+          if (j == nvalues)
+            return thrust::make_tuple(index_x, count_x + count_y);
+          return thrust::make_tuple(index_y, count_x + count_y);
+        });
+    auto iter_tuple = new_end.second.get_iterator_tuple();
+    auto new_sz =
+        thrust::distance(mx_rows_indices.begin(), thrust::get<0>(iter_tuple));
+    mx_rows_indices.resize(new_sz);
+    count_set.resize(new_sz);
+    mx_rows.resize(new_sz);
+    if (LogSoftMax) {
+      log_softmax_max_buffer.resize(new_sz * nvalues);
+      auto log_softmax_max_buffer_ptr = log_softmax_max_buffer.data().get();
+      for (int64_t index = 0; index < new_sz; index++) {
+        auto current = mx_rows_indices[index];
+        cudaMemcpy(
+            log_softmax_max_buffer_ptr + index * nvalues,
+            values.data_ptr<scalar_t>() + current * nvalues,
+            nvalues * sizeof(scalar_t),
+            cudaMemcpyDeviceToDevice);
```

Comment:
This is incorrect for `nvalues>1`.


(i) The corresponding code has to be equivalent to https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/sparse/SoftMax.cpp#L330-L335

(ii) Consider the following example:
```
indices = [[0, 0, 0, 1, 1], [0, 3, 4, 1, 3]]
s = 100
values = [[11*s, -11*s], [14*s, -14*s], [15*s, -15*s], [22*s, -22*s], [24*s, -24*s]]

for device in ['cpu', 'cuda']:
    X = torch.sparse_coo_tensor(indices, values, (2, 5, 2), dtype=float, device=device)
    Y = torch.sparse.log_softmax(X, dim=0)
    print(Y)
```
that outputs
```
tensor(indices=tensor([[0, 0, 0, 1, 1],
                       [0, 3, 4, 1, 3]]),
       values=tensor([[    0.,     0.],
                      [-1000.,     0.],
                      [    0.,     0.],
                      [    0.,     0.],
                      [    0., -1000.]]),
       size=(2, 5, 2), nnz=5, dtype=torch.float64, layout=torch.sparse_coo)
tensor(indices=tensor([[0, 0, 0, 1, 1],
                       [0, 3, 4, 1, 3]]),
       values=tensor([[    0.,     0.],
                      [-1000.,   -inf],
                      [    0.,     0.],
                      [    0.,     0.],
                      [    0.,   -inf]]),
       device='cuda:0', size=(2, 5, 2), nnz=5, dtype=torch.float64,
       layout=torch.sparse_coo)
```
where the result from cuda target is wrong. Try `s=10` to get matching results.

(iii) When dumping the content of `log_softmax_max_buffer`, it is
```
log_softmax_max_buffer=[1100 -1100 2200 -2200 2400 -2400 1500 -1500 ]
```
but the output should be
```
log_softmax_max_buffer=[1100 -1100 2200 -2200 2400 -1400 1500 -1500 ]
```

Reply:
Thanks for notice me about this case. The example was really helpful to understand the error in the code. I updated the way the max values are computed.